internal/querycoordv2/meta/resource_manager.go

```diff
@@ -0,0 +1,600 @@
+}
+
+func (rm *ResourceManager) ContainResourceGroup(rgName string) bool {
+	rm.rwmutex.RLock()
+	defer rm.rwmutex.RUnlock()
+	return rm.groups[rgName] != nil
+}
+
+func (rm *ResourceManager) GetResourceGroup(rgName string) (*ResourceGroup, error) {
+	rm.rwmutex.RLock()
+	defer rm.rwmutex.RUnlock()
+
+	if rm.groups[rgName] == nil {
+		return nil, ErrRGNotExist
+	}
+
+	rm.checkRGNodeStatus(rgName)
+	return rm.groups[rgName], nil
+}
+
+func (rm *ResourceManager) ListResourceGroups() []string {
+	rm.rwmutex.RLock()
+	defer rm.rwmutex.RUnlock()
+
+	return lo.Keys(rm.groups)
+}
+
+func (rm *ResourceManager) FindResourceGroupByNode(node int64) (string, error) {
+	rm.rwmutex.RLock()
+	defer rm.rwmutex.RUnlock()
+
+	return rm.findResourceGroupByNode(node)
+}
+
+func (rm *ResourceManager) findResourceGroupByNode(node int64) (string, error) {
+	for name, group := range rm.groups {
+		if group.containsNode(node) {
+			return name, nil
+		}
+	}
+
+	return "", ErrNodeNotAssignToRG
+}
+
+func (rm *ResourceManager) HandleNodeUp(node int64) (string, error) {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+
+	if rm.nodeMgr.Get(node) == nil {
+		return "", ErrNodeNotExist
+	}
+
+	if ok, _ := rm.nodeMgr.IsStoppingNode(node); ok {
+		return "", ErrNodeStopped
+	}
+
+	// if node already assign to rg
+	rgName, err := rm.findResourceGroupByNode(node)
+	if err == nil {
+		log.Info("HandleNodeUp: node already assign to resource group",
+			zap.String("rgName", rgName),
+			zap.Int64("node", node),
+		)
+		return rgName, nil
+	}
+
+	// add new node to default rg
+	rm.groups[DefaultResourceGroupName].handleNodeUp(node)
+	log.Info("HandleNodeUp: assign node to default resource group",
+		zap.String("rgName", DefaultResourceGroupName),
+		zap.Int64("node", node),
+	)
+	return DefaultResourceGroupName, nil
+}
+
+func (rm *ResourceManager) HandleNodeDown(node int64) (string, error) {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+
+	if rm.nodeMgr.Get(node) == nil {
+		return "", ErrNodeNotExist
+	}
+
+	rgName, err := rm.findResourceGroupByNode(node)
+	if err == nil {
+		log.Info("HandleNodeDown: remove node from resource group",
+			zap.String("rgName", rgName),
+			zap.Int64("node", node),
+		)
+		return rgName, rm.groups[rgName].handleNodeDown(node)
+	}
+
+	return "", ErrNodeNotAssignToRG
+}
+
+func (rm *ResourceManager) TransferNode(from, to string) error {
+	rm.rwmutex.Lock()
+	defer rm.rwmutex.Unlock()
+
+	if rm.groups[from] == nil || rm.groups[to] == nil {
+		return ErrRGNotExist
+	}
+
+	if len(rm.groups[from].nodes) == 0 {
+		return ErrRGIsEmpty
+	}
+
+	rm.checkRGNodeStatus(from)
+	//todo: a better way to choose a node with least balance cost
+	node := rm.groups[from].GetNodes()[0]
+	err := rm.unassignNode(from, node)
+	if err != nil {
+		// interrupt transfer, unreachable logic path
+		return err
+	}
+
+	rm.checkRGNodeStatus(to)
+	err = rm.assignNode(to, node)
+	if err != nil {
+		// roll back, unreachable logic path
+		rm.assignNode(from, node)
+	}
+
```

Comment:
if error occurs here, querycoord may disconnect from etcd, assignNode might also fail. In this case, one node will went missing.